safe/gui/tools/minimum_needs/needs_calculator_dialog.py

```diff
@@ -61,148 +96,107 @@ def __init__(self, parent=None):
-        # see #2206
-        ok_button = self.button_box.button(QtGui.QDialogButtonBox.Ok)
-        if not has_fields:
-            ok_button.setEnabled(False)
-        else:
-            ok_button.setEnabled(True)
+        # create memory layer
+        output_layer = (
+            '%s_minimum_needs' % input_layer.name())
+        output_layer = create_memory_layer(
+            output_layer,
+            input_layer.geometryType(),
+            input_layer.crs(),
+            input_layer.fields())
+
+        # monkey patching input layer to make it work with
+        # prepare vector layer function
+        temp_layer = input_layer
+        temp_layer.keywords = {
+            'layer_purpose': layer_purpose_aggregation['key']}
+
+        # add keywords to output layer
+        copy_layer(temp_layer, output_layer)
+
+        return output_layer
 
     def accept(self):
         """Process the layer and field and generate a new layer.
 
         .. note:: This is called on OK click.
 
         """
-        index = self.cboFields.currentIndex()
-        field_name = self.cboFields.itemData(
-            index, QtCore.Qt.UserRole)
+        # Create a new layer for output layer
+        output_layer = self.prepare_new_layer(self.layer.currentLayer())
 
-        index = self.cboPolygonLayers.currentIndex()
-        layer_id = self.cboPolygonLayers.itemData(
-            index, QtCore.Qt.UserRole)
-        # noinspection PyArgumentList
-        layer = QgsMapLayerRegistry.instance().mapLayer(layer_id)
+        # Monkey patching output layer to make it work with
+        # minimum needs calculator
+        output_layer.keywords['layer_purpose'] = (
+            layer_purpose_aggregation['key'])
+        output_layer.keywords['inasafe_fields'] = (
+            {displaced_field['key']: self.displaced.currentField(),
+             aggregation_name_field['key']:
+                 self.aggregation_name.currentField()})
 
-        file_name = layer.source()
+        # remove unnecessary fields & rename inasafe fields
+        rename_remove_inasafe_fields(output_layer)
 
         try:
-            output_layer = self.minimum_needs(input_layer, field_name)
-        except ValueError:
+            self.minimum_needs(output_layer)
+        except Exception as e:
+            LOGGER.debug(e)
             return
 
-        new_file = file_name[:-4] + '_perka7' + '.shp'
+        # remove monkey patching keywords
+        del output_layer.keywords
+
+        # write layer to geopackage file
+        settings = QSettings()
+        default_user_directory = settings.value(
+            'inasafe/defaultUserDirectory', defaultValue='')
+
+        if default_user_directory:
+            path = os.path.join(default_user_directory, output_layer.name())
+            if not os.path.exists(path):
+                os.makedirs(path)
+            data_store = Folder(path)
+            data_store.add_layer(output_layer, output_layer.name())
```

Comment:
I think you need to remove this line. Because the layer will be added twice with the line 195